### DISCUSSION TITLE
Request method propagation

### DISCUSSION THREAD

**Discussion by m4tthumphrey** [dedoc/scramble#602] [Ideas]
Hello, great library, I am really enjoying using it!

I love the "Method calls on a request object" functionality however it doesn't lend well to reusability, that is `$request->get()` has to be called explicitly within the route scope (ie controller method etc). I am trying to reuse as much of the pagination code in my app as possible as per the example below.

As you can see, I only need to call `$request->integer('page.size', self::DEFAULT_SIZE)` and `$request->integer('page.number', 1)` once for every route that I want to use this functionality from this trait. This does not seem to compatible with Scramble.

Now I do not know enough about Reflection in PHP (yet) but I'm wondering if this could evolve into a feature or if it simple isn't possible? If its the latter I can move on swiftly but if there is some chance it could work eventually, I would be fine to repeat myself until such a time it's available.

Thoughts?

```php
$paginated = $this->paginated($venues, $request, ['id', 'name', 'address', 'description', 'coordinates', 'img_stadium', 'img_categories', 'destination_id']);
```

```php
trait Paginatable
{
    const MAX_SIZE     = 100;
    const DEFAULT_SIZE = 10;

    public function limit(int $size, $default = self::DEFAULT_SIZE): int
    {
        if ($size && $size <= self::MAX_SIZE && $size > 0) {
            return $size;
        }

        return $default;
    }

    public function paginated(Builder $collection, Request $request, array $fields)
    {
        return $collection
            ->paginate(
                $this->limit($request->integer('page.size', self::DEFAULT_SIZE)),
                $fields,
                'page[number]',
                $request->integer('page.number', 1)
            )
            ->withPath($this->cleanUrl())
        ;
    }
}

**Comment by romalytvynenko** [comment]
@m4tthumphrey Can you please show some examples of your real route's methods and what is your expected ideal API documentation? 

I kind of get what you're asking about, but not 100% sure

**Comment by m4tthumphrey** [comment]
@romalytvynenko  

Thanks for the reply! 

Controller looks like this. The doc should mention query params:

`page.size`
`page.number`
`country`

```php
    public function venues(Request $request): AnonymousResourceCollection
    {
        $countryId = $request->integer('country');

        $venues = Venue::with('destination:id,name,iata_code,country_id')
            ->where('name', '<>','')
            ->when($countryId, fn ($query) => $query->whereHas('destination', fn ($query) => $query->whereCountryId($countryId)));

        $paginated = $this->paginated($venues, $request, ['id', 'name', 'address', 'description', 'coordinates', 'img_stadium', 'img_categories', 'destination_id']);

        return VenueResource::collection($paginated);
    }

**Reply by romalytvynenko** [reply]
@m4tthumphrey awesome, thanks!

The first idea that comes to mind mind is to enable you to do something like this (this is not implemented yet!):

```php
#[Param(name: 'page.size', type: 'int', default: self::DEFAULT_SIZE)]
#[Param(name: 'page.number', type: 'int', default: 1)]
public function paginated(Builder $collection, Request $request, array $fields)
{
    // ...
}
```

Usually this sort of attributes are applied to the controller methods. But in this case it would be also cool if Scramble supported these annotations on methods called in a controller's methods.

Technically it is possible to infer these types when the argument is passed like this into other function, but it needs to be carefully planned. And such annotations for more complex cases like yours seems fine to me.

What do you think?

**Reply by m4tthumphrey** [reply]
Yes this could definitely work! Does this project use annotations at all at this point? I haven’t seen anywhere yet, only phpdoc.

Could this also work outside of a controller? Ie 

A) My example above is from a trait 
B) How about if it wasn’t a trait but say a manager or a service class? Could annotations be used there too? Ie any method anyway in the project that a controller method calls?

**Reply by romalytvynenko** [reply]
> Does this project use annotations at all at this point?

In the latest version I've added few ones (https://scramble.dedoc.co/developers/reference#attributes), I also plan to add more to give the ability to do a manual takeover when needed.

> Could this also work outside of a controller?

Yeah, the idea right now is that if any method annotated this way is called in a controller, Scramble will add parameters to the documentation.

**Reply by m4tthumphrey** [reply]
Awesome! I look forward to the next release! Thanks!

**Reply by romalytvynenko** [reply]
@m4tthumphrey I'm not sure when exactly I'd release it 🫣

**Reply by m4tthumphrey** [reply]
Haha sorry yes I know, i just meant whenever it is released!

**Comment by m4tthumphrey** [comment]
@romalytvynenko 

Please let me know if a need to create an issue for this but constants do not seem to work when using 

`$request->integer('page.size', self::DEFAULT_VALUE)`

It just treats it as blank. I think this code is handled [vendor/dedoc/scramble/src/Support/Type/Literal/LiteralIntegerType.php](vendor/dedoc/scramble/src/Support/Type/Literal/LiteralIntegerType.php) but as I said before I do not know enough bout reflection, yet.

**Reply by m4tthumphrey** [reply]
@romalytvynenko I've tried at length to figure out how to get the const value but (without hardcoding class names!) I cannot figure it out :(
